pkg/shim/z_shim_cmd.go

```diff
@@ -4,80 +4,115 @@
 )
 
 const (
 	// SubCommand is the first argument to the clone of the nomad agent process
 	// for invoking the exec2 driver sandbox shim.
 	SubCommand = "exec2-shim"
 
-	// deadline is the total amount of time we allow this process to be alive;
-	// in this case we just invoke landlock and then exec into another
-	// process
-	deadline = 1 * time.Second
+	// ExitWrongArgs indicates the shim has terminated early due to recieving
+	// the wrong expected arguments. We use a special return code here since logs
+	// will not have been configured yet.
+	ExitWrongArgs = 40
+
+	// ExitBadLogging indicates the shim has terminated early due to being unable
+	// to open stdout or stderr output files (fifos).
+	ExitBadLogging = 41
 )
 
 // init is the entrypoint for the 'nomad e2e-shim' invocation of nomad
 //
 // The argument format is as follows,
 //
-// 1. nomad            <- the executable name
-// 2. exec2-shim       <- this subcommand
-// 3. true/false       <- include default unveil paths
-// 4. [mode:path, ...] <- list of additional unveil paths
-// 5. --               <- sentinel between following commands
+// 0. nomad            <- the executable name
+// 1. exec2-shim       <- this subcommand
+// 2. true/false       <- include default unveil paths
+// 3. <stdout path>    <- path to named pipe for standard output
+// 4. <stderr path>    <- path to named pipe for standard error
+// 5. [mode:path, ...] <- list of additional unveil paths
+// 6. --               <- sentinel between following commands
 func init() {
 	subproc.Do(SubCommand, func() int {
-		ctx, cancel := subproc.Context(deadline)
-		defer cancel()
-
-		// ensure we die in a timely manner
-		subproc.SetExpiration(ctx)
+		// we need to ignore the stop signal (which is sent to the entire
+		// process group) so that we stay alive and can capture the exit code
+		// of the child task process
+		sigs := make(chan os.Signal, 1)
+		signal.Notify(sigs)
+		go func() {
+			<-sigs // do nothing; say alive
+		}()
 
 		if n := len(os.Args); n <= 4 {
 			subproc.Print("failed to invoke e2e-shim with sufficient args: %d", n)
-			return subproc.ExitFailure
+			return ExitWrongArgs
 		}
 
 		// get the unveil paths and the rest of the command(s) to run
 		// from our command arguments
-		args := os.Args[3:] // chop off 'nomad exec2-shim <defaults>'
+		args := os.Args[5:] // chop off 'nomad exec2-shim <defaults>'
 		defaults := os.Args[2] == "true"
+		outPipePath := os.Args[3]
+		errPipePath := os.Args[4]
 		paths, commands := split(args)
+		paths = append(paths, "w:"+outPipePath)
+		paths = append(paths, "w:"+errPipePath)
+
+		stdout, stderr, err := util.OpenPipes(outPipePath, errPipePath)
+		if err != nil {
+			subproc.Print("failed to open output pipes: %v", err)
+			return ExitBadLogging
+		}
+
+		// give ourselves a way to write to the stderr pipe for printing fatal errors
+		debug := func(format string, args ...any) {
+			_, _ = io.WriteString(stderr, fmt.Sprintf(format, args...))
+		}
```

Comment:
Just for my clarity, this it the same named pipe we eventually give to the workload as its stderr? So these logs will show up in `nomad alloc logs`?

Reply:
I recently was chatting with @schmichael (IIRC) about trying to pull the `executor.log` for the exec v1 driver into the alloc logs too! So cool that we're already doing it here, and I think this reduces my worry that we could be exposing something sensitive in the logs this way.